### PR TITLE
Add new module manageiq_policies

### PR DESCRIPTION
**Pull request by yaacov**
**SUMMARY**

ManageIQ is an open source management platform for Hybrid IT.

This change is adding:

    manageiq_policies module, responsible for policy_profiles management in ManageIQ

All the modules, including docs, tests and usage examples can be found here

Currently, the only requirement for the modules is manageiq-api-client-python

**ISSUE TYPE**

    New Module Pull Request

**COMPONENT NAME**

manageiq_policies.py (module)

**ANSIBLE VERSION**

ansible 2.5.0.0

### CODE REVIEW

**Comment by yaacov** [comment]
@gtanzillo @abellotti (ManageIQ team), @zgalor @cben @simon3z (Containers Management team) @joelddiaz (Ops team) - please review.

**Comment by yaacov** [comment]
shipit

**Comment by joelddiaz** [comment]
@yaacov when running with state 'absent' against a policy that doesn't exist it errors out. I would expect the module to run successfully saying something like 'policy ABC already not associated with object' rather than stop the playbook:

```
TASK [absent non-existing policy] *******************************************************************************************************************************
task path: /home/jdiaz/ansible/miq-test.yml:299
Using module file /home/jdiaz/tmp/ansible-provider/lib/ansible/modules/remote_management/manageiq/manageiq_policies.py
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "manageiq_connection": {
                "ca_bundle_path": null, 
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "token": null, 
                "url": "REDACTED", 
                "username": "admin", 
                "verify_ssl": true
            }, 
            "policy_profiles": [
                {
                    "name": "nonexistent profile"
                }
            ], 
            "resource_name": "cm-ops", 
            "resource_type": "provider", 
            "state": "absent"
        }
    }, 
    "msg": "policy_profiles where {'name': 'nonexistent profile'} does not exist in manageiq"
}
```

File: lib/ansible/modules/remote_management/manageiq/manageiq_policies.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    default: null
+  resource_name:
+    description:
+      - the relevant resource name in manageiq
+    required: true
+    default: null
+'''
+
+EXAMPLES = '''
+- name: Create new policy_profile for a provider in ManageIQ
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_policies.py]: s/create new/assign/

File: lib/ansible/modules/remote_management/manageiq/manageiq_policies.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    resource_type: 'provider'
+    policy_profiles:
+      - name: openscap profile
+    manageiq_connection:
+      url: 'http://127.0.0.1:3000'
+      username: 'admin'
+      password: 'smartvm'
+      verify_ssl: False
+
+- name: Remove a policy_profile for a provider in ManageIQ
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_policies.py]: s/remove/unassign/

File: lib/ansible/modules/remote_management/manageiq/manageiq_policies.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    default: 'present'
+  policy_profiles:
+    description:
+      - list of dictionaries, each includes the policy_profile 'name' key.
+      - required if state is present or absent.
+    required: false
+    default: null
+  resource_type:
+    description:
+      - the relevant resource type in manageiq
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_policies.py]: "the relevant" not clear enough IMHO.  something like "the type of the resource to which the profile should be [un]assigned" ?
same for `resource_name` below.

File: lib/ansible/modules/remote_management/manageiq/manageiq_policies.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        self.resource_url = '{api_url}/{resource_type}/{resource_id}'.format(
+            api_url=self.api_url,
+            resource_type=resource_type,
+            resource_id=resource_id)
+
+    def query_profile_href(self, profile):
+        """ Add or Update the policy_profile href field
+
+        Example:
+            [{name: STR}, ...] => [{name: STR, href: STR}, ...]
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_policies.py]: this func takes and returns single dict, not array of dicts.

**Review by cben — COMMENTED** [review]
still reviewing, some nitpicks, overall great!

File: lib/ansible/modules/remote_management/manageiq/manageiq_policies.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+            response = self.client.post(url, action=action, resources=profiles_to_post)
+        except Exception as e:
+            msg = "Failed to {action} profile: {error}".format(
+                action=action,
+                error=e)
+            self.module.fail_json(msg=msg)
+
+        # check all entities in result to be successfull
+        for result in response['results']:
+            if not result['success']:
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_policies.py:257]: doesn't the underlying client raise if there is any `result['error']`?
hmm, I see, this post returns a`"results"` list, and the client `_result_processor` doesn't peek into that.
https://github.com/mkoura/manageiq-api-client-python/blob/master/src/manageiq_client/api.py#L69
(it does only if you use its `Action` class)
=> LGTM, I'll see about improving that in client...

**Review by cben — APPROVED** [review]
:+1:

**Comment by yaacov** [comment]
shipit

**Comment by cben** [comment]
shitit

**Comment by cben** [comment]
shipit

**Comment by cben** [comment]
bot_status

**Comment by yaacov** [comment]
> @yaacov when running with state 'absent' against a policy that doesn't exist it errors out. I would expect the module to run successfully saying something like 'policy ABC already not associated with object' rather than stop the playbook:

@joelddiaz fixed 🎉 , please review.

**Comment by joelddiaz** [comment]
@yaacov lgtm with your recent changes

**Comment by yaacov** [comment]
Sorry everyone, bad rebase.

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
@zgalor @cben @joelddiaz @elad661 #31774 changed the way we use `manageiq_connection`, last commit updates this PR for the new `manageiq_connection` var.

please re-shipit

**Comment by yaacov** [comment]
ready_for_review

**Comment by yaacov** [comment]
-label aws 
-label azure 
-label c:inventory/contrib_script 
-label cli/adhoc 
-label cli/console 
-label cli/doc 
-label cli/pull 
-label cloud 
-label cloudstack 
-label constants 
-label docker 
-label docs_pull_request 
-label errors/ 
-label executor/play_iterator 
-label executor/task_executor 
-label executor/task_queue_manager 
-label executor/task_result 
-label f5 
-label inventory 
-label inventory/group 
-label inventory/host 
-label module_utils/azure_rm_common 
-label module_utils/basic 
-label module_utils/facts 
-label module_utils/powershell 
-label module_utils/vmware 
-label net_tools 
-label networking 
-label openstack 
-label ovirt 
-label support:core 
-label vmware

**Comment by zgalor** [comment]
shipit

**Comment by yaacov** [comment]
bot_status

**Comment by yaacov** [comment]
-label cli/adhoc
-label cli/console
-label cli/doc
-label cli/pull
-label cloudstack
-label constants
-label docs_pull_request
-label errors/
-label executor/play_iterator
-label executor/task_executor
-label executor/task_queue_manager
-label executor/task_result
-label f5
-label inventory
-label inventory/group
-label inventory/host
-label module_utils/azure_rm_common
-label module_utils/basic
-label module_utils/facts
-label module_utils/powershell
-label module_utils/vmware
-label net_tools
-label ovirt
-label support:core

**Comment by yaacov** [comment]
Remove `find_collection_resource_or_fail` method, see #32588

shipit

@cben @zgalor @elad661 please re-shipit

**Comment by cben** [comment]
shipit

**Comment by zgalor** [comment]
shipit

**Comment by yaacov** [comment]
@ryansb do you know who can help remove the incorrect labels this pull request collected due to my bad rebase ?

**Comment by yaacov** [comment]
Cloned to https://github.com/ansible/ansible/pull/32892

Closing this one, because of the bad labelling.

**Pull request closed without merging**